### PR TITLE
[doc] Improve Manual Optimization Example

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR provides a complete GAN example with `manual optimization` as requested by some users.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/common/optimizers.rst [hunk 1]
```diff
@@ -67,27 +67,91 @@ Here is the same example as above using a ``closure``.
...
+        # true data from generated ones
+        return -1 * (tr.log(d_x).mean() + tr.log(1 - d_z).mean())
+
+    def sample_z(self, n) -> Tensor:
+        sample = self._Z.sample((n,))
+        return sample
+
+    def sample_G(self, n) -> Tensor:
+        z = self.sample_z(n)
+        return self.G(z)
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:119]: Is it worth having all this? A reader still wouldnt be able to run it as `MnistDenseGenerator` is not available

File: docs/source/common/optimizers.rst [hunk 2]
```diff
@@ -67,27 +67,91 @@ Here is the same example as above using a ``closure``.
...
+        X, _ = batch
+        batch_size = X.shape[0]
+        g_X = self.sample_G(batch_size)
+        d_z = self.D(g_X)
+        g_loss = self.generator_loss(d_z)
+
+        # zero_grad should be called before manual_backward
+        g_opt.zero_grad()
+        self.manual_backward(g_loss)
+
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:135]:
```suggestion
```

File: docs/source/common/optimizers.rst [hunk 3]
```diff
@@ -67,27 +67,91 @@ Here is the same example as above using a ``closure``.
...
+        d_z = self.D(g_X.detach())
+
+        d_loss = self.discriminator_loss(d_x, d_z)
+
+        # zero_grad should be called before manual_backward
+        d_opt.zero_grad()
+        self.manual_backward(d_loss)
+        d_opt.step()
+
+        self.log_dict({'g_loss': g_loss, 'd_loss': d_loss}, prog_bar=True, logger=True)
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:149]: Simplify

```suggestion
        self.log_dict({'g_loss': g_loss, 'd_loss': d_loss})
```

File: docs/source/common/optimizers.rst [hunk 4]
```diff
@@ -67,27 +67,91 @@ Here is the same example as above using a ``closure``.
...
-        loss_gen = self.compute_generator_loss(...)
-
-        # zero_grad needs to be called before backward
-        opt_gen.zero_grad()
-        self.manual_backward(loss_gen)
-        opt_gen.step()
 
-        # compute discriminator loss
-        loss_dis = self.compute_discriminator_loss(...)
+.. code-block:: python
```
Comment (Borda) [inline docs/source/common/optimizers.rst:73]:
```suggestion
.. testcode::
```

File: docs/source/common/optimizers.rst (hunk 1 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst:119]: Good catch ! Removing it.

File: docs/source/common/optimizers.rst (hunk 4 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst:73]: Not sure it makes sense to have test code if the code contains ... for some part right ?

Comment (Borda) [reply docs/source/common/optimizers.rst:73]: contains what? a not long time ago we had PRs with fixing python syntax in docs and these tests do not cost anything and catches these kinds of issues

File: docs/source/common/optimizers.rst [hunk 5]
```diff
@@ -100,36 +163,64 @@ Toggling means that all parameters from B exclusive to A will have their ``requi
 When performing gradient accumulation, there is no need to perform grad synchronization during the accumulation phase.
 Setting ``sync_grad`` to ``False`` will block this synchronization and improve your training speed.
 
-Here is an example on how to use it:
+
+Here is an example for advanced use-case.
+
 
 .. code-block:: python
```
Comment (Borda) [inline docs/source/common/optimizers.rst:170]:
```suggestion
.. testcode::
```
drop the `...` bellow

File: docs/source/common/optimizers.rst [hunk 6]
```diff
@@ -67,28 +67,91 @@ Here is the same example as above using a ``closure``.
...
+            # Important: This property activate ``manual optimization`` for this model
+            return False
+
+        def generator_loss(self, d_z: Tensor) -> Tensor:
+            # the closer ``d_z`` is from 1,
+            # the better the generator is able to fool the discriminator
+            return -1 * tr.log(d_z).mean()
+
+        def discriminator_loss(self, d_x: Tensor, d_z: Tensor) -> Tensor:
+            # the closer is ``d_x`` from 1 and ``dz`` from 0,
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:97]:
```suggestion
            # the closer is ``d_x`` from 1 and ``d_z`` from 0,
```

File: docs/source/common/optimizers.rst [hunk 7]
```diff
@@ -67,28 +67,91 @@ Here is the same example as above using a ``closure``.
...
+
+        @property
+        def automatic_optimization(self):
+            # Important: This property activate ``manual optimization`` for this model
+            return False
+
+        def generator_loss(self, d_z: Tensor) -> Tensor:
+            # the closer ``d_z`` is from 1,
+            # the better the generator is able to fool the discriminator
+            return -1 * tr.log(d_z).mean()
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:94]:
```suggestion
            return -1 * torch.log(d_z).mean()
```

File: docs/source/common/optimizers.rst [hunk 8]
```diff
@@ -67,28 +67,91 @@ Here is the same example as above using a ``closure``.
...
+
+        def discriminator_loss(self, d_x: Tensor, d_z: Tensor) -> Tensor:
+            # the closer is ``d_x`` from 1 and ``dz`` from 0,
+            # the better the discriminator is able to distinguish
+            # true data from generated ones
+            return -1 * (tr.log(d_x).mean() + tr.log(1 - d_z).mean())
```
Comment (akihironitta) [inline docs/source/common/optimizers.rst:100]:
```suggestion
            return -1 * (torch.log(d_x).mean() + torch.log(1 - d_z).mean())
```

File: docs/source/common/optimizers.rst [hunk 9]
```diff
@@ -67,28 +67,91 @@ Here is the same example as above using a ``closure``.
...
+
+        def __init__(self):
+            super().__init__()
+            self.G = Generator(...)
+            self.D = Discriminator(...)
+
+        @property
+        def automatic_optimization(self):
+            # Important: This property activate ``manual optimization`` for this model
+            return False
```
[the remark this answers is not in the document]
Comment (tchaton) [reply docs/source/common/optimizers.rst:89]: Better to expose the property directly in the doc even if it was the same behaviour.

File: docs/source/common/optimizers.rst (hunk 6 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst]: Done !

File: docs/source/common/optimizers.rst (hunk 5 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst:170]: Done !

File: docs/source/common/optimizers.rst (hunk 7 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst:94]: Done !

File: docs/source/common/optimizers.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/common/optimizers.rst:89]: Actually, the correct way is setting `self.automatic_optimization` in the `__init__` as we have a getter/settter for it which modifies `self._automatic_optimization`

https://github.com/PyTorchLightning/pytorch-lightning/blob/484dce11ecb92aab13c29d26b57cbf8c7522ceeb/pytorch_lightning/core/lightning.py#L167-L180

So @akihironitta is correct

[the remark this answers is not in the document]
Comment (tchaton) [reply docs/source/common/optimizers.rst:89]: I missed the setter, thanks !

File: docs/source/common/optimizers.rst [hunk 10]
```diff
@@ -21,10 +21,21 @@ Manual optimization
 For advanced research topics like reinforcement learning, sparse coding, or GAN research, it may be desirable
 to manually manage the optimization process. To do so, do the following:
 
-* Override your LightningModule ``automatic_optimization`` property to return ``False``
+* Set the ``automatic_optimization`` property to ``False`` in your ``LightningModule`` ``__init__`` function
 * Drop or ignore the optimizer_idx argument
 * Use ``self.manual_backward(loss)`` instead of ``loss.backward()``.
 
+.. testcode:: python
```
Comment (pierresegonne) [inline docs/source/common/optimizers.rst:28]: Nice addition!

I don't know if it's doable, but putting that code snippet below the right bullet might improve understandability

File: docs/source/common/optimizers.rst [hunk 11]
```diff
@@ -67,28 +78,89 @@ Here is the same example as above using a ``closure``.
...
+            sample = self._Z.sample((n,))
+            return sample
+
+        def sample_G(self, n) -> Tensor:
+            z = self.sample_z(n)
+            return self.G(z)
+
+        def training_step(self, batch, batch_idx, optimizer_idx, *args):
+            # Implementation follows https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html
+            g_opt, d_opt = self.optimizers()
```
Comment (pierresegonne) [inline docs/source/common/optimizers.rst:121]: Maybe little comment about the unused optimiser_idx arg?

Comment (tchaton) [reply docs/source/common/optimizers.rst:121]: Removed entirely.

File: docs/source/common/optimizers.rst (hunk 10 above)
Comment (tchaton) [reply docs/source/common/optimizers.rst:28]: Good idea !

**Review by akihironitta — APPROVED** [review]

File: docs/source/common/optimizers.rst [hunk 12]
```diff
@@ -21,10 +21,32 @@ Manual optimization
...
+            # Important: This property activate ``manual optimization`` for your model
+            self.automatic_optimization = False
+
+.. testcode:: python
+
+    from pytorch_lightning import LightningModule
+
+    class MyModel(LightningModule):
+
+        def training_step(batch, batch_idx):
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:44]: why split?

```suggestion
            self.automatic_optimization = False

        def training_step(batch, batch_idx):
```

File: docs/source/common/optimizers.rst [hunk 13]
```diff
@@ -67,28 +89,89 @@ Here is the same example as above using a ``closure``.
...
+        def generator_loss(self, d_z: Tensor) -> Tensor:
+            # the closer ``d_z`` is from 1,
+            # the better the generator is able to fool the discriminator
+            return -1 * torch.log(d_z).mean()
+
+        def discriminator_loss(self, d_x: Tensor, d_z: Tensor) -> Tensor:
+            # the closer is ``d_x`` from 1 and ``d_z`` from 0,
+            # the better the discriminator is able to distinguish
+            # true data from generated ones
+            return -1 * (torch.log(d_x).mean() + torch.log(1 - d_z).mean())
```
Comment (carmocca) [inline docs/source/common/optimizers.rst:120]: This is unused in the code

```suggestion
```

**Review by Borda — APPROVED** [review]

**Review by ananyahjha93 — APPROVED** [review]
